pkgs/development/compilers/llvm/common/libcxx/default.nix

```diff
@@ -13,7 +13,7 @@
 , python3
 , fixDarwinDylibNames
 , version
-, cxxabi ? if stdenv.hostPlatform.isFreeBSD then libcxxrt else null
+, cxxabi ? null
```

Comment:
not sure if it matters, but to get `libcxxrt` working correctly i had to add `-DLIBCXX_ENABLE_NEW_DELETE_DEFINITIONS=ON` to the cxxCMakeFlags  https://libcxx.llvm.org/BuildingLibcxx.html#using-libcxxrt-on-linux to get freebsd cross and linux pkgsLLVM to work for various packages that used c++.

Reply:
I can't speak for cross, but I was just able to compile gmp-with-cxx on a freebsd machine using the stdenv derived from these changes.

Reply:
i was under the impression that libcxxrt was required for freebsd https://github.com/NixOS/nixpkgs/pull/211230 tho if it isn't it would be nice to get rid of it at some point -- it's been a minor pain point when working on libcxx (i tried to remove it a few months ago).

Reply:
yeah, it seems like for whatever reason it's not required native (I double checked by compiling boost just now). Someone should take a look at cross though...

Reply:
cross builds work fine using `libcxxabi` for all available boost versions
```bash
for x in {168..185} '' 16x 17x 18x; do nix-build -A pkgsCross.x86_64-freebsd.boost$x; done
```
only failures were ones aliased to throw or missing versions.

Reply:
I'm not sure what you're referring to - is "with libcxxabi" the consequence of my diff?

Reply:
> I'm not sure what you're referring to - is "with libcxxabi" the consequence of my diff?

yes. if `cxxabi == null` you'll get `libcxxabi`.

Reply:
Doesn't normal FreeBSD use libcxxrt?  Are we sure we'd want to be different?

Reply:
> Doesn't normal FreeBSD use libcxxrt? Are we sure we'd want to be different?

one reason to keep it is to support binaries compiled on vanilla freebsd that expect libcxxrt -- we wouldn't have to add any additional logic when binary patching.

i can't imagine there is code trying to dlopen libcxxrt, but that doesn't mean it doesn't exist.

removing libcxxrt (and external cxxabi) from the libcxx derivation reduces complexity and maintenance burden so i'd be happy to see it removed.

if it stays i needs to add `-DLIBCXX_ENABLE_NEW_DELETE_DEFINITIONS=ON`. freebsd has used the libcxx `new.cpp` since llvm12, i think. `libcxxrt` doesn't provide `new [] / delete []` (or some variation of) which breaks linux llvm17+ when building against it, and c++ programs don't build (duckdb) on freebsd cross without it (tho i thought duckdb worked via cross a few months ago).

Reply:
It seems like boost started working with libcxxabi when cross compiling to FreeBSD with the LLVM 12 -> 16 update, so maybe it was just a bug in libcxxabi all along that got fixed?  I'm fine with unconditionally using libcxxabi, as long as we bear in mind that we might find reasons in the future that we need to match what normal FreeBSD does and switch back to libcxxrt.

Reply:
so if freebsd is going to use libcxxabi then all the libcxxrt code can be removed as it is now dead. here is the diff to remove it against staging-next. tested `nix-build -A tests.cc-wrapper.llvmTests` and `nix-build -A pkgsCross.x86_64-freebsd.llvmPackages_17.libcxx`.

<details><summary>drop-libcxxrt.diff</summary>

```diff
diff --git a/pkgs/development/compilers/llvm/common/libcxx/default.nix b/pkgs/development/compilers/llvm/common/libcxx/default.nix
index 5b4c2ca56ae1..346a6f6e3fcb 100644
--- a/pkgs/development/compilers/llvm/common/libcxx/default.nix
+++ b/pkgs/development/compilers/llvm/common/libcxx/default.nix
@@ -8,27 +8,17 @@
 , runCommand
 , substitute
 , cmake
-, lndir
 , ninja
 , python3
 , fixDarwinDylibNames
 , version
-, cxxabi ? if stdenv.hostPlatform.isFreeBSD then libcxxrt else null
-, libcxxrt
 , libunwind
 , enableShared ? !stdenv.hostPlatform.isStatic
 }:
 
-# external cxxabi is not supported on Darwin as the build will not link libcxx
-# properly and not re-export the cxxabi symbols into libcxx
-# https://github.com/NixOS/nixpkgs/issues/166205
-# https://github.com/NixOS/nixpkgs/issues/269548
-assert cxxabi == null || !stdenv.hostPlatform.isDarwin;
 let
   basename = "libcxx";
   pname = basename;
-  cxxabiName = "lib${if cxxabi == null then "cxxabi" else cxxabi.libName}";
-  runtimes = [ "libcxx" ] ++ lib.optional (cxxabi == null) "libcxxabi";
 
   # Note: useLLVM is likely false for Darwin but true under pkgsLLVM
   useLLVM = stdenv.hostPlatform.useLLVM or false;
@@ -46,9 +36,8 @@ let
       cp -r ${monorepoSrc}/third-party "$out"
     '') + ''
       cp -r ${monorepoSrc}/runtimes "$out"
-    '' + (lib.optionalString (cxxabi == null) ''
       cp -r ${monorepoSrc}/libcxxabi "$out"
-    '')) else src;
+    '') else src;
 
   cxxabiCMakeFlags = lib.optionals (lib.versionAtLeast release_version "18") [
     "-DLIBCXXABI_USE_LLVM_UNWINDER=OFF"
@@ -66,15 +55,13 @@ let
   ];
 
   cxxCMakeFlags = [
-    "-DLIBCXX_CXX_ABI=${cxxabiName}"
-  ] ++ lib.optionals (cxxabi == null && lib.versionAtLeast release_version "16") [
+    "-DLIBCXX_CXX_ABI=libcxxabi"
+  ] ++ lib.optionals (lib.versionAtLeast release_version "16") [
     # Note: llvm < 16 doesn't support this flag (or it's broken); handled in postInstall instead.
     # Include libc++abi symbols within libc++.a for static linking libc++;
     # dynamic linking includes them through libc++.so being a linker script
     # which includes both shared objects.
     "-DLIBCXX_STATICALLY_LINK_ABI_IN_STATIC_LIBRARY=ON"
-  ] ++ lib.optionals (cxxabi != null) [
-    "-DLIBCXX_CXX_ABI_INCLUDE_PATHS=${lib.getDev cxxabi}/include"
   ] ++ lib.optionals (stdenv.hostPlatform.isMusl || stdenv.hostPlatform.isWasi) [
     "-DLIBCXX_HAS_MUSL_LIBC=1"
   ] ++ lib.optionals (lib.versionAtLeast release_version "18" && !useLLVM && stdenv.hostPlatform.libc == "glibc" && !stdenv.hostPlatform.isStatic) [
@@ -92,7 +79,7 @@ let
   ];
 
   cmakeFlags = [
-    "-DLLVM_ENABLE_RUNTIMES=${lib.concatStringsSep ";" runtimes}"
+    "-DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi"
   ] ++ lib.optionals (useLLVM && !stdenv.hostPlatform.isWasm) [
     # libcxxabi's CMake looks as though it treats -nostdlib++ as implying -nostdlib,
     # but that does not appear to be the case for example when building
@@ -104,7 +91,7 @@ let
     "-DCMAKE_CXX_COMPILER_WORKS=ON"
     "-DUNIX=ON" # Required otherwise libc++ fails to detect the correct linker
   ] ++ cxxCMakeFlags
-    ++ lib.optionals (cxxabi == null) cxxabiCMakeFlags;
+    ++ cxxabiCMakeFlags;
 
 in
 
@@ -120,19 +107,15 @@ stdenv.mkDerivation (rec {
   '';
 
   nativeBuildInputs = [ cmake ninja python3 ]
-    ++ lib.optional stdenv.isDarwin fixDarwinDylibNames
-    ++ lib.optional (cxxabi != null) lndir;
-
-  buildInputs = [ cxxabi ]
-    ++ lib.optionals (useLLVM && !stdenv.hostPlatform.isWasm) [ libunwind ];
-
-  # libc++.so is a linker script which expands to multiple libraries,
-  # libc++.so.1 and libc++abi.so or the external cxxabi. ld-wrapper doesn't
-  # support linker scripts so the external cxxabi needs to be symlinked in
-  postInstall = lib.optionalString (cxxabi != null) ''
-    lndir ${lib.getDev cxxabi}/include $dev/include/c++/v1
-    lndir ${lib.getLib cxxabi}/lib $out/lib
-    libcxxabi=$out/lib/lib${cxxabi.libName}.a
+    ++ lib.optional stdenv.isDarwin fixDarwinDylibNames;
+
+  buildInputs = lib.optionals (useLLVM && !stdenv.hostPlatform.isWasm) [ libunwind ];
+
+  # vanilla FreeBSD uses libcxxrt to provide the cxxabi functionality, where
+  # nixpkgs is using the libc++abi bundled with LLVM.
+  postInstall = lib.optionalString (stdenv.hostPlatform.isFreeBSD) ''
+    ${lib.optionalString enableShared ''ln -s libc++abi.so $out/lib/libcxxrt.so''}
+    ln -s libc++abi.a $out/lib/libcxxrt.a
   ''
   # LIBCXX_STATICALLY_LINK_ABI_IN_STATIC_LIBRARY=ON doesn't work for LLVM < 16 or
   # external cxxabi libraries so merge libc++abi.a into libc++.a ourselves.
@@ -140,13 +123,12 @@ stdenv.mkDerivation (rec {
   # GNU binutils emits objects in LIFO order in MRI scripts so after the merge
   # the objects are in reversed order so a second MRI script is required so the
   # objects in the archive are listed in proper order (libc++.a, libc++abi.a)
-  + lib.optionalString (cxxabi != null || lib.versionOlder release_version "16") ''
-    libcxxabi=''${libcxxabi-$out/lib/libc++abi.a}
-    if [[ -f $out/lib/libc++.a && -e $libcxxabi ]]; then
+  + lib.optionalString (lib.versionOlder release_version "16") ''
+    if [[ -f $out/lib/libc++.a && -f $out/lib/libc++abi.a ]]; then
       $AR -M <<MRI
         create $out/lib/libc++.a
         addlib $out/lib/libc++.a
-        addlib $libcxxabi
+        addlib $out/lib/libc++abi.a
         save
         end
     MRI
```
</details>

Reply:
Can this be a followup PR?